librz/core/rop.c

```diff
@@ -0,0 +1,1238 @@
+// SPDX-FileCopyrightText: 2024 z3phyr <giridh1337@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_core.h>
+#include <rz_asm.h>
+#include <rz_util/rz_log.h>
+#include <rz_util/rz_regex.h>
+#include <rz_rop.h>
+
+static bool is_end_gadget(const RzAnalysisOp *aop, const ut8 crop) {
+	if (aop->family == RZ_ANALYSIS_OP_FAMILY_SECURITY) {
+		return false;
+	}
+	switch (aop->type) {
+	case RZ_ANALYSIS_OP_TYPE_TRAP:
+	case RZ_ANALYSIS_OP_TYPE_RET:
+	case RZ_ANALYSIS_OP_TYPE_UCALL:
+	case RZ_ANALYSIS_OP_TYPE_RCALL:
+	case RZ_ANALYSIS_OP_TYPE_ICALL:
+	case RZ_ANALYSIS_OP_TYPE_IRCALL:
+	case RZ_ANALYSIS_OP_TYPE_UJMP:
+	case RZ_ANALYSIS_OP_TYPE_RJMP:
+	case RZ_ANALYSIS_OP_TYPE_IJMP:
+	case RZ_ANALYSIS_OP_TYPE_IRJMP:
+	case RZ_ANALYSIS_OP_TYPE_JMP:
+	case RZ_ANALYSIS_OP_TYPE_CALL:
+		return true;
+	default:
+		return false;
+	}
+
+	if (!crop) {
+		return false;
+	}
+	switch (aop->type) {
+	case RZ_ANALYSIS_OP_TYPE_CJMP:
+	case RZ_ANALYSIS_OP_TYPE_UCJMP:
+	case RZ_ANALYSIS_OP_TYPE_CCALL:
+	case RZ_ANALYSIS_OP_TYPE_UCCALL:
+	case RZ_ANALYSIS_OP_TYPE_CRET:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static int rz_rop_process_asm_op(const RzCore *core, const RzCoreAsmHit *hit, RzAsmOp *asmop, RzAnalysisOp *aop, unsigned int *size, char **asmop_str, char **asmop_hex_str) {
+	ut8 *buf = malloc(hit->len);
+	if (!buf) {
+		return -1;
+	}
+	if (rz_io_nread_at(core->io, hit->addr, buf, hit->len) < 0) {
+		free(buf);
+		return -1;
+	}
+	rz_asm_set_pc(core->rasm, hit->addr);
+	if (rz_asm_disassemble(core->rasm, asmop, buf, hit->len) < 0) {
+		free(buf);
+		return -1;
+	}
+	rz_analysis_op_init(aop);
+	rz_analysis_op(core->analysis, aop, hit->addr, buf, hit->len, RZ_ANALYSIS_OP_MASK_DISASM);
+	*size += hit->len;
+
+	// Append assembly operation string
+	if (asmop_str) {
+		*asmop_str = rz_str_append(*asmop_str, rz_asm_op_get_asm(asmop));
+		*asmop_str = rz_str_append(*asmop_str, "; ");
+	}
+
+	// Append hex string of assembly operation
+	if (asmop_hex_str) {
+		char *asmop_hex = rz_asm_op_get_hex(asmop);
+		*asmop_hex_str = rz_str_append(*asmop_hex_str, asmop_hex);
+		free(asmop_hex);
+	}
+
+	free(buf);
+	return 0;
+}
+
+static int rz_rop_print_table_mode(const RzCore *core, const RzCoreAsmHit *hit, const RzList /*<RzCoreAsmHit *>*/ *hitlist,
```

Comment:
This one could be converted to returning `bool` since only two values are possible